File: scripts/list_microphones.py
#!/usr/bin/env python3
"""
List available microphones
"""

import pyaudio

p = pyaudio.PyAudio()

print("🎙️  Available Microphones:")
print("=" * 50)

for i in range(p.get_device_count()):
    info = p.get_device_info_by_index(i)
    if info['maxInputChannels'] > 0:
        print(f"\n[{i}] {info['name']}")
        print(f"    Sample Rate: {int(info['defaultSampleRate'])} Hz")
        print(f"    Channels: {info['maxInputChannels']}")

p.terminate()

print("\n" + "=" * 50)
print("To use a specific mic, update the Python scripts")
print("Add: input_device_index=X")
print("where X is the number in brackets")

File: src/list-microphones.ts
import { spawn } from "child_process";

export interface Microphone {
  index: number;
  name: string;
  sampleRate: number;
  channels: number;
}

export async function listMicrophones(): Promise<Microphone[]> {
  return new Promise((resolve, reject) => {
    const process = spawn("uvx", [
      "--with",
      "pyaudio",
      "python3",
      "-c",
      `
import pyaudio
import json

p = pyaudio.PyAudio()
devices = []

for i in range(p.get_device_count()):
    info = p.get_device_info_by_index(i)
    if info['maxInputChannels'] > 0:
        devices.append({
            'index': i,
            'name': info['name'],
            'sampleRate': int(info['defaultSampleRate']),
            'channels': info['maxInputChannels']
        })

print(json.dumps(devices))
p.terminate()
      `,
    ]);

    let output = "";
    process.stdout.on("data", (data) => {
      output += data.toString();
    });

    process.on("close", (code) => {
      if (code === 0) {
        try {
          const devices = JSON.parse(output.trim());
          resolve(devices);
        } catch (e) {
          reject(new Error("Failed to parse microphone list"));
        }
      } else {
        reject(new Error(`Failed to list microphones: exit code ${code}`));
      }
    });
  });
}
